### PR TITLE
Adding documentation for using valueFrom

### PR DESCRIPTION
**Pull request by coreydaley**
Dependent on https://github.com/openshift/origin/pull/14143

### CODE REVIEW

**Comment by coreydaley** [comment]
@bparees @bmcelvee ptal

File: _topic_map.yml [hunk 1]
```diff
@@ -701,6 +701,8 @@ Topics:
         File: build_output
       - Name: Build Strategy Options
         File: build_strategies
+      - Name: BuildConfig Environment
```
Comment (bparees) [inline _topic_map.yml]: s/BuildConfig Environment/Build Environment/  (pretty much everywhere)

File: dev_guide/builds/buildconfig_environment.adoc [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+[NOTE]
+====
+You can also manage environment variables defined in the `BuildConfig` with the
+xref:../../dev_guide/environment_variables.adoc#dev-guide-environment-variables[`oc set env`] command.
+====
+
+[[setting-environment-variables]]
+== Setting Environment Variables
+
```
Comment (bparees) [inline dev_guide/builds/buildconfig_environment.adoc]: this needs some sort of context/introductory text.

Comment (bparees) [inline dev_guide/builds/buildconfig_environment.adoc]: also build_loglevel is a problematic example.  pick something less special, eg "examplevariable=examplevalue"

File: dev_guide/builds/buildconfig_environment.adoc [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+<2> The value to assign to the environment variable
+
+
+[[referencing-other-environment-variables]]
+== Referencing Other Environment Variables
+
+You can reference the value of a previously defined environment
+variable by using the `$()` syntax. If the environment variable
+reference can not be resolved, the value will be set to the provided
+string as shown below.
```
Comment (bparees) [inline dev_guide/builds/buildconfig_environment.adoc]: it doesn't seem like we should have to be documenting this.  is this behavior (which is general k8s env var behavior) not already covered elsewhere in the docs?  if not, i would propose that it should be and we shouldn't have to say anything about it specific to builds.

File: dev_guide/builds/buildconfig_environment.adoc [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  - name: SECOND_ENV
+    value: $(FIRST_ENV) <1>
+----
+
+<1> `$(FIRST_ENV)` is expanded to the value of `FIRST_ENV` which is `myvalue` unless `FIRST_ENV` does not exist, in which case the value will be `$(FIRST_ENV)`.
+
+[[escaping-environment-variable-references]]
+== Escaping Environment Variable References
+
+You can escape an environment variable reference by using a double dollar sign.
```
Comment (bparees) [inline dev_guide/builds/buildconfig_environment.adoc]: same thing here..this isn't build specific behavior.  we're just finally supporting what everything else already did.

File: dev_guide/builds/buildconfig_environment.adoc [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+[[using-valuefrom-references]]
+== Using valueFrom References
+
+[[using-value-and-valuefrom-together]]
+=== Using value and valueFrom Together
+
+If both `value` and `valueFrom` are supplied for a build environment variable
+definition, `value` will be used and the `valueFrom` will be ignored.  in the following
+example, the value of `MY_ENV` would be `myvalue` regardless of what the `valueFrom` would
+resolve to.
```
Comment (bparees) [inline dev_guide/builds/buildconfig_environment.adoc]: and same thing here, except that we are a little specialized in terms of what references we do/do not support, which is worthy of a note somewhere in the builds doc.

File: dev_guide/builds/buildconfig_environment.adoc [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+[source,yaml]
+----
+env:
+  - name: MY_ENV
+    value: myvalue
+    valueFrom:
+      secretKeyRef:
+        name: mysecret
+        key: username
+----
```
Comment (bparees) [inline dev_guide/builds/buildconfig_environment.adoc]: it seems a bit silly to have this example repeated immediately below.  i'd say this section (which is just an intro to the concept of valuefrom) maybe doesn't need a specific example, given the subsections provide examples.

(but i also think all this content should be refactored to be generic product doc on the behavior of envvars, not build specific, aside from whatever special cases we have in builds)

**Review by bparees — CHANGES_REQUESTED** [review]

File: dev_guide/builds/buildconfig_environment.adoc (hunk 2 above)
Comment (coreydaley) [inline dev_guide/builds/buildconfig_environment.adoc]: updated in next push

File: _topic_map.yml (hunk 1 above)
Comment (coreydaley) [inline _topic_map.yml]: Updated in next push

**Comment by bmcelvee** [comment]
@coreydaley thanks for the rebase! @bparees are you still reviewing this PR? 

Thanks!

File: dev_guide/builds/build_strategies.adoc [hunk 7]
```diff
@@ -568,31 +577,52 @@ jenkinsPipelineStrategy:
...
+The `BuildConfig` values take precedence.
+* If you start via `oc start-build -e`, the values for the environment variables specified in the `-e` option take precedence.  And if you
+specify an environment variable not listed in the `BuildConfig`, they will be added as a Jenkins job parameter definitions. And, again, any
+changes you make from the Jenkins console to the parameters corresponding to the environment variables are ignored.  The `BuildConfig`
+and what you specify via `oc start-build -e` takes precedence.
+* If you start the Jenkins job via the Jenkins console, then you can control the setting of the parameters via the Jenkins console as part
+of starting a build for the job.
+
+Please visit the xref:../../dev_guide/builds/build_environment.adoc#dev-guide-build-build-environment[build environment]
+section for more advanced topics.
```
Comment (bparees) [inline dev_guide/builds/build_strategies.adoc]: this all looks redundant to the content immediately below it?  (everything except the new link to the advanced topics anyway).

also s/topics/usage/ here.

File: dev_guide/downward_api.adoc [hunk 8]
```diff
@@ -293,8 +292,8 @@ field to describe the desired resources that correspond to the
 `*spec.resources*` field:
 +
 ====
+[source,yaml]
 ----
-....
```
Comment (bparees) [inline dev_guide/downward_api.adoc]: this is here for a reason, it indicates this is not a fully complete pod definition.

File: dev_guide/downward_api.adoc [hunk 9]
```diff
@@ -331,18 +330,163 @@ spec:
             resourceFieldRef:
               containerName: client-container
               resource: requests.memory
-....
```
Comment (bparees) [inline dev_guide/downward_api.adoc]: leave this here or complete the example

File: dev_guide/downward_api.adoc (hunk 8 above)
Comment (bparees) [inline dev_guide/downward_api.adoc]: so either it should be left, or this should be fleshed out as a fully creatable example.

**Review by bparees — CHANGES_REQUESTED** [review]

**Comment by bparees** [comment]
@bmcelvee yes, sorry. just did another round.

File: dev_guide/builds/build_environment.adoc [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+<2> the following `fieldPath` refers to a jsonpath in the `build`
+<3> the jsonpath to the field in the `build`
+
+[[using-container-resources-as-environment-variables]]
+=== Using Container Resources as Environment Variables
+
+Referencing container resources using valueFrom in build environment variables is currently not supported
+as the references are resolved before the container is created.
+
+
```
Comment (coreydaley) [inline dev_guide/builds/build_environment.adoc]: I had previously removed it via your comment: https://github.com/openshift/openshift-docs/pull/4557#discussion_r121723411

File: dev_guide/downward_api.adoc (hunk 9 above)
Comment (coreydaley) [inline dev_guide/downward_api.adoc]: updated in next push

File: dev_guide/downward_api.adoc (hunk 8 above)
Comment (coreydaley) [inline dev_guide/downward_api.adoc]: updated in next push

File: dev_guide/builds/build_environment.adoc [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+== Overview
+
+[NOTE]
+====
+You can also manage environment variables defined in the `BuildConfig` with the
+xref:../../dev_guide/environment_variables.adoc#dev-guide-environment-variables[`oc set env`] command.
+====
+
+[[using-valuefrom-references]]
+== Using valueFrom References
```
Comment (coreydaley) [inline dev_guide/builds/build_environment.adoc]: updated in next push

File: dev_guide/builds/build_strategies.adoc (hunk 7 above)
Comment (coreydaley) [inline dev_guide/builds/build_strategies.adoc]: updated in next push

File: dev_guide/builds/build_strategies.adoc [hunk 12]
```diff
@@ -492,6 +498,9 @@ xref:../../install_config/http_proxies.adoc#configuring-hosts-for-proxies-using-
 global build settings using Ansible].
 endif::[]
 
+Please visit the xref:../../dev_guide/builds/build_environment.adoc#dev-guide-build-build-environment[build environment]
+section for more advanced topics.
```
Comment (coreydaley) [inline dev_guide/builds/build_strategies.adoc]: updated in next push

File: dev_guide/builds/build_strategies.adoc [hunk 13]
```diff
@@ -304,6 +307,9 @@ xref:../../install_config/http_proxies.adoc#configuring-hosts-for-proxies-using-
...
```
Comment (coreydaley) [inline dev_guide/builds/build_strategies.adoc]: updated in next push

File: dev_guide/builds/build_strategies.adoc [hunk 14]
```diff
@@ -156,6 +156,9 @@ sourceStrategy:
       value: "true"
 ----
 
+Please visit the xref:../../dev_guide/builds/build_environment.adoc#dev-guide-build-build-environment[build environment]
+section for more advanced topics.
```
Comment (coreydaley) [inline dev_guide/builds/build_strategies.adoc]: updated in next push

File: dev_guide/builds/build_environment.adoc (hunk 10 above)
Comment (coreydaley) [inline dev_guide/builds/build_environment.adoc]: updated in next push

File: dev_guide/builds/build_environment.adoc [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+:experimental:
+:toc: macro
+:toc-title:
+:prewrap!:
+
+toc::[]
+
+
+[[overview]]
+== Overview
```
Comment (coreydaley) [inline dev_guide/builds/build_environment.adoc]: updated

File: dev_guide/builds/build_environment.adoc (hunk 10 above)
Comment (coreydaley) [inline dev_guide/builds/build_environment.adoc]: Added to the Downward API page

**Comment by coreydaley** [comment]
@bparees ptal

File: dev_guide/builds/build_environment.adoc [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+----
+
+<1> the value should be resolved using the following references
+<2> the following `fieldPath` refers to a jsonpath in the `build`
+<3> the jsonpath to the field in the `build`
+
+[[using-container-resources-as-environment-variables]]
+== Using Container Resources as Environment Variables
+
+Referencing container resources using valueFrom in build environment variables is currently not supported
```
Comment (bparees) [inline dev_guide/builds/build_environment.adoc]: s/currently//

File: dev_guide/builds/build_strategies.adoc [hunk 17]
```diff
@@ -568,31 +577,28 @@ jenkinsPipelineStrategy:
...
+The `BuildConfig` values take precedence.
+* If you start via `oc start-build -e`, the values for the environment variables specified in the `-e` option take precedence.  And if you
+specify an environment variable not listed in the `BuildConfig`, they will be added as a Jenkins job parameter definitions. And, again, any
+changes you make from the Jenkins console to the parameters corresponding to the environment variables are ignored.  The `BuildConfig`
+and what you specify via `oc start-build -e` takes precedence.
+* If you start the Jenkins job via the Jenkins console, then you can control the setting of the parameters via the Jenkins console as part
+of starting a build for the job.
+
+Please visit the xref:../../dev_guide/builds/build_environment.adoc#dev-guide-build-build-environment[build environment]
+section for more advanced usage.
```
Comment (bparees) [inline dev_guide/builds/build_strategies.adoc]: @gabemontero please review this revised content (just this pipeline env section specifically)

Comment (bparees) [inline dev_guide/builds/build_strategies.adoc]: i'm not sure moving all of the content into a true "note" annotation make sense.  it's an awfully long note :)

**Comment by bparees** [comment]
pending my one nit and @gabemontero's sign off on the pipeline env content, this lgtm.

**Comment by coreydaley** [comment]
@bparees updated

**Review by gabemontero — APPROVED** [review]
pipeline env specifics look correct

**Comment by coreydaley** [comment]
@bmcelvee ptal, I think this one is finally ready!

**Review by bparees — APPROVED** [review]

File: dev_guide/builds/build_strategies.adoc (hunk 17 above)
Comment (bmcelvee) [inline dev_guide/builds/build_strategies.adoc]: I agree, it's a bit too much for a note. :) I'll merge what we have here and open a follow-up PR.

**Comment by bmcelvee** [comment]
[rev_history]
|xref:../dev_guide/builds/build_environment.adoc#dev-guide-build-build-environment[Build Environment]
|Added xref:../dev_guide/builds/build_environment.adoc#dev-guide-build-build-environment[Build Environment] chapter.
%
|xref:../dev_guide/builds/build_strategies.adoc#dev-guide-build-strategy-options[Build Strategy Options]
|Added links pointing to the xref:../dev_guide/builds/build_environment.adoc#dev-guide-build-build-environment[Build Environment] chapter, and added content on mapping between `BuildConfig` environment variables and Jenkins job parameters to the xref:../dev_guide/builds/build_strategies.adoc#jenkins-pipeline-strategy-environment[Environment Variables] section.
%
|xref:../dev_guide/downward_api.adoc#dev-guide-downward-api[Downward API]
|Updated `.json` references to `.yaml`, and added valueFrom content to the xref:../dev_guide/downward_api.adoc#dev-guide-downward-api[Downward API] chapter.
%

**Pull request merged**